doc/OnlineDocs/contributed_packages/edi/formulation.rst

```diff
@@ -0,0 +1,233 @@
+    :language: python 
+    :dedent: 8
+    :start-after: # BEGIN: Formulation_Snippet_08
+    :end-before: # END: Formulation_Snippet_08
+
+We recommend that most users should be using the ``f.ConstraintList()`` function, with the ``f.Constraint()`` function being reserved for under-the-hood usage.
+
+
+Declaring Runtime (Black-Box) Constraints
+-----------------------------------------
+
+See the :doc:`Runtime (Black-Box) Constraints <./blackboxconstraints>` Documentation
+
+One of the main features of EDI is the streamlined implementation of Black-Box Constraints.  A *Black-Box* is defined as a routine that performs hidden computation not visible EDI, pyomo, or more generally the optimization algorithm.  However, it is **not** assumed that black-boxes are unable to return gradient information.  A black-box in this context may be capable of returning arbitrary derivative information.
+
+Black-box constraints are considered to be a sub-class of a more general class of constraints called *Runtime Constraints*, that is constraints that are not actually constructed until the optimization routine is actually running.  In most cases, Runtime Constraints are approximated as linear by the solver, and therefore a Runtime Constraint is expected to provide function evaluations and gradient information.
+
+The use of Runtime constraints requires a black box model that is discussed in detail in the dedicated documentation (see :doc:`here <./blackboxconstraints>`), but for the purposes of demonstrating the constructors, a simple black box example will appear in all of the code snippets below.
+
+Runtime Constraints are declared one of two ways, just as regular constraints.  The ``f.RuntimeConstraint()`` constructor is available: 
+
+.. literalinclude:: ../../../../pyomo/contrib/edi/tests/test_docSnippets.py
+    :language: python 
+    :dedent: 8
+    :start-after: # BEGIN: Formulation_Snippet_09
+    :end-before: # END: Formulation_Snippet_09
+
+
+The ``f.RuntimeConstraint()`` constructor takes in the following inputs:
+
+.. py:function:: f.RuntimeConstraint(outputs, operators, inputs, black_box)
+    :noindex:
+
+    Declares a runtime constraint in a pyomo.edi.formulation
+
+    :param outputs: The outputs of the black box function
+    :type outputs: pyomo.environ.Var or list or tuple
+    :param operators: The operators that are used to construct constraints.  Currently, only equality constraints are supported and will be the default no matter what is passed in here (see `this issue <https://github.com/codykarcher/pyomo/issues/4>`__)
+    :type operators: str or list or tuple
+    :param inputs: The inputs to the black box function
+    :type inputs: pyomo.environ.Var or list or tuple
+    :param black_box: The object that stores the black-box function.   See the :doc:`black box constraint documentation <./blackboxconstraints>` for details on constructing this object
+    :type black_box: pyomo.contrib.edi.BlackBoxFunctionModel
```

Comment:
I really hate docstrings in code.  It's a personal preference, and it's something I go back and forth on as to if I should just suck it up and adopt.  But for now, my preferred path is to provide good documentation (and potentially some fancy tricks for a `help` command in EDI), instead of doing inline comments.  I could possibly be persuaded, but for now, (politely) rejecting this suggestion.